nomad/eval_broker.go

```diff
@@ -158,6 +163,8 @@ func NewEvalBroker(ctx context.Context, timeout, initialNackDelay, subsequentNac
 		timeWait:             make(map[string]*time.Timer),
 		initialNackDelay:     initialNackDelay,
 		subsequentNackDelay:  subsequentNackDelay,
+		enqueuedTime:         make(map[string]time.Time),
+		dequeuedTime:         make(map[string]time.Time),
```

Comment:
We know in advance that these maps are going to get populated and that will cause memory allocations in the hot path of the eval broker. Maybe we should pre-allocate some nice round number like 256.

Also, we're creating these empty maps in `NewEvalBroker` which gets called on every server, including followers. But when there's a leader election the new leader goes through the state store and re-enqueues all the evaluations into its own eval broker. We should make sure we're clearing these maps when the follower steps down. Otherwise if the leaders change and then change back, we'll end up with very weird metrics.

Reply:
Yeah, I think we need to make a more conscious decision on how to handle leadership transition. 

Pulling my comment from the internal RFC discussion:

> But looking at the the leadership transition code, when we restore evals we just call the same Enqueue() method, so the wait times are likely to go down and service time metrics dropped because we don't have the dequeue time anymore.
> 
> Ignoring the restored evals (meaning, don't emit these new metrics for them) can cause a gap until evals created after the leadership transition start to flow in.
> 
> Between the two options, I think a gap may be better than incorrect values?

Maybe a good way to handle would be:

1. Clean new timing maps on leadership step down.
2. Don't set enqueue time for restored evals on new leader. We may need to create a separate method for eval restore, like this:
    ```go
	func (b *EvalBroker) Restore(eval *structs.Evaluation) {
		b.l.Lock()
		defer b.l.Unlock()
		b.processEnqueue(eval, "", false)
	}
	
	func (b *EvalBroker) Enqueue(eval *structs.Evaluation) {
		b.l.Lock()
		defer b.l.Unlock()
		b.processEnqueue(eval, "", true)
	}
	
	func (b *EvalBroker) processEnqueue(eval *structs.Evaluation, token string, trackTime bool) {
		// ...
	}
    ```

I'm still not 100% sure with the right approach here, so it would be good to test a few options and document the intention and decision (in code, documentation, and in the RFC).

Reply:
> We know in advance that these maps are going to get populated and that will cause memory allocations in the hot path of the eval broker. Maybe we should pre-allocate some nice round number like 256.

good thinking. Added in [383b9ce](https://github.com/hashicorp/nomad/pull/20329/commits/383b9ce2e776199fd3d62db4998ae9c1ca1bb17e)